### ISSUE TITLE
Unable to create non-initramfs UKI which boots

### ISSUE THREAD

**Issue by Winterhuman** [systemd/mkosi#2067]
`mkosi` version: Git main (14/11/23)
`uname`: Linux 6.6.1-arch1-1 #1 SMP PREEMPT_DYNAMIC x86_64 GNU/Linux

# Description

I've been trying to create something along the lines of https://github.com/orgs/uapi-group/discussions/32, however, so far I've not been successful in making a UKI that fully boots.

Currently, it just reaches "Welcome to Arch Linux", and hitting enter as prompted does nothing, and the logs end somewhere at the sound target, network name lookup, or Virtual Console {Starting,Stopping,etc} (it varies between runs).

The output works completely fine with `mkosi qemu`, which doesn't appear to start the firstboot prompt, however, booting the UKI on real hardware fails as described above.

# Configs

Here's the relevant config, there are no other configs or filled directories outside of `mkosi.pkgmgrtree`, which is a simple `/etc/pacman.d` drop-in:

```
[Distribution]
Distribution=arch
# For setting 'ParallelDownloads = 10' as a drop-in
PackageManagerTrees=mkosi.pkgmgrtree

[Content]
MakeInitrd=no
Bootable=no
KernelCommandLine=systemd.volatile=overlay
Packages=
	bash
	linux
	util-linux
	systemd

[Output]
Format=uki
CompressOutput=yes
ImageId=UkiOS
```

# Command

`bin/mkosi` (presence of `-i` makes no difference)

# Logs

[mkosi.log](https://github.com/systemd/mkosi/files/13355290/mkosi.log)

**Comment by poettering** [comment]
I recently played around with the concept, and it worked fine. See: https://github.com/poettering/diskomator/

**Comment by Winterhuman** [comment]
I'm noticing a few differences between our configs:

- You set output format to "esp" instead of "uki".
- You use Fedora instead of Arch Linux for the distribution.
- You include `linux-firmware` as a package.

That last one is difficult for me to test because the image becomes 459.5MiB in size from including it, which is too large for my ESP (I couldn't get `qemu` to not complain about getty, so I can't do that currently), which is strange given you claim your final image is ~300MiB.

**Comment by poettering** [comment]
Fedora splits up the firmware package btw. Some of the worst offenders are not included I think.

**Comment by Winterhuman** [comment]
Ah, that's probably why then. Hmm, that makes things tricky...

It's weird that the firstboot service seems to be the problem, it just awkwardly prints its prompt in the logs, but the logs don't stop until a decent bit afterwards, and I'm starting to wonder if the reason hitting enter isn't working is because my keyboard isn't being picked up (perhaps being related to `linux-firmware` not being present).

**Comment by Winterhuman** [comment]
Update: I've managed to get `mkosi qemu` working, and it... works completely fine? The firstboot prompt never happens when run with `qemu`, and I get prompted for login details as expected, however, booting the same image on real hardware *does* come up with the firstboot prompt and stops the whole process

**Comment by behrmann** [comment]
Did you change anything with the above config?

**Comment by Winterhuman** [comment]
I added the following to the config to get `mkosi qemu` working:

```
[Host]
QemuMem=3G
QemuFirmware=uefi
#QemuKernel=/home/hyper/Public/OSmaker/UKI/mkosi.output/ukios.efi
```

And I added `mkosi.rootpw` since I realised I forgot. However, neither changed the booting issue on real hardware

**Comment by poettering** [comment]
note that diskomator specifies a credential to set the TZ on the kernel cmdline, so that firstboot doesn't query you. did you see that?

**Comment by poettering** [comment]
you could alternatively just symlink UTC to /etc/timezone

**Comment by Winterhuman** [comment]
I'm never able to get past the "Hit any key to continue" part of the firstboot prompt, so I'm not sure. Running `timedatectl` in the qemu session shows my current timezone being set

**Comment by Winterhuman** [comment]
Just tried your idea by running `ln -sf` inside a `mkosi.finalize` script. Now when booting, it says the firstboot service starts and finishes successfully (the "Welcome to Arch Linux" text never appears this time), however, the log messages abruptly stopped at "Starting user manager" and doesn't continue further.

So, that rules out the firstboot prompt holding everything up from trying to read my input, something else causes systemd to get stuck at seemingly random targets & points which doesn't happen in `qemu`.

**Comment by poettering** [comment]
(btw, is this on intel graphics? if the firmware files aren't available the intel drm driver freezes the machine hard. I had trouble with that on a physical machine, which was simply fixed by adding the intel firmware to the image)

**Comment by Winterhuman** [comment]
Nope, AMD dedicated graphics. Though now that you've said that, I realise `mesa` or any graphic drivers aren't included in the packages, so I'll try that now

**Comment by Winterhuman** [comment]
Okay, I managed to reach the `archlinux login:` part of the boot process, but then the screen froze and the underscore wasn't flashing. It seems weird why I'd need graphics drivers to get the TTY working, or why `mesa` alone doesn't fix this, but at least we're closer to an answer

**Comment by behrmann** [comment]
@Winterhuman Assuming that this is your current `mkosi.conf`
```
[Distribution]
Distribution=arch

[Content]
MakeInitrd=no
Bootable=no
KernelCommandLine=systemd.volatile=overlay
Packages=
	bash
	linux
	util-linux
	systemd

[Output]
Format=uki
CompressOutput=yes
ImageId=UkiOS

[Host]
QemuMem=3G
QemuFirmware=uefi
```

This builds and boots to prompt fine for me on Arch with current main. The only other thing is a `mkosi.rootpw`. I didn't add a `mkosi.finalize` and there's no firstboot stuff happening for me.

**Comment by Winterhuman** [comment]
Closing issue...

It *was* `linux-firmware`, that's why @behrmann and `mkosi qemu` couldn't reproduce the issue. I was only able to test it now because I figured out how to slim down the package far enough so the UKI would fit in my ESP, which I achieved by removing everything under `/usr/lib/firmware/` that wasn't listed in my `dracut`-generated UKIs `lsinitrd` output.

> `/usr/lib/modules/$(uname -r)/` could be reduced too, which is also what `dracut` does, but thankfully, it wasn't necessary to shrink the image to 291MiB.

It's annoying that Arch's `linux-firmware` package is so massive, but, I can't imagine there's anything the `mkosi` project could do about it. Thanks you all for your help! (and sorry for the wild goose chase)
